### ISSUE TITLE
gcc: compilation optimization march/mtune level

### ISSUE THREAD

**Issue by yesimon** [bioconda/bioconda-recipes#2354]
I'm opening this issue so that we can standardize on instruction level support. Typically compilation with gcc will default to `-march=x86-64` which is the most generic x86 support available and should work on all processors. However, newer processors have newer instruction sets such as SSSE3, AVX, etc that are only supported on a subset of processors.

Some packages have Makefiles that include parameters such as `-march=native`, which optimizes for the current architecture but may only work on the cpu level that the package was compiled on (in our case whatever Travis provides), but this software wouldn't work on older cpus.

One approach would be to standardize on a level such as `-march=x86-64` - this precludes optimization in favor of compatibility and is how OS system packages are built. For OS X, Homebrew standardizes on `-march=core2` which I think is around the time that Apple switched to Intel so that level of support should be universal.

Another approach may be to add yet more to the build string suffix. In this case, there can be suffixes for differing levels of CPU  such as `sandybridge` or `skylake` https://gcc.gnu.org/onlinedocs/gcc-6.2.0/gcc/x86-Options.html#x86-Options

**Comment by kyleabeauchamp** [comment]
We may wish to check in with conda-forge and Continuum about best practices.

Clearly a lot of bioinformatics can benefit from having some level of optimization so it would be great if we had systems in place to maximize performance.

**Comment by yesimon** [comment]
Another approach, such as by RAxML is to compile multiple binaries with suffixes in a single package: 
https://github.com/bioconda/bioconda-recipes/blob/master/recipes/raxml/build.sh

**Comment by kyleabeauchamp** [comment]
As food for thought, intel has a whitepaper that claims having AVX instructions may improve BWA:

http://www.intel.com/content/www/us/en/healthcare-it/solutions/documents/performance-benchmarking-of-burrow-wheeler-aligner.html

I'm not sure if their results apply to the current versions of BWA, but I for one would be thrilled to have faster alignment.

**Comment by kyleabeauchamp** [comment]
Looks like this might be in a non-standard branch of bwa, sadly: https://sourceforge.net/p/bio-bwa/mailman/message/34618535/

**Comment by yesimon** [comment]
Yes the whitepaper is very misleading by because it claims bwa was compiled with gcc flags `-O2 -funroll-loops`, which should still default to `-march=generic`. It's also comparing two completely different CPUs.

I have actually tested compiling with `icc` with ipo optimizations and even then the performance difference was negligible (0-5% improvement).

**Comment by blaiseli** [comment]
I have just learned about bioconda and conda today, and it looks like it could be a good substitute for the set of install scripts I started to use in order to be able to install programs at the user level without having to bother the admins of our computing server.

I'm by no means a C/C++ expert, but I took the habit to try to build from source as much as possible and add the `-march=native  -fomit-frame-pointer` compile flags, hoping that this would make the programs more efficient.

I don't really understand what would be the issue using this kind of approach in bioconda recipes (at least for programs that are not distributed as binaries).
Is conda/bioconda meant to be used to build programs on one machine and then distribute them on others with potentially different processors ?

**Comment by yesimon** [comment]
@blaiseli Yes conda is meant to be a platform for portable packages (it should work on any x86-64) CPU. Otherwise compiling with the latest instruction sets may make binaries invalid and fail even on relatively recent hardware.

For your own software feel free to compile from source and try to optimize as much as possible. Back in the olden days Gentoo was popular because you had the opportunity to compile the entire OS from source with `-march=native`. It quickly lost out in popularity to other distros with binary packages because compiling from source is difficult, takes a long time, often didn't benefit from native optimizations, and CPUs because fast enough that these optimizations were unimportant.

However scientific software often does include optimization for advanced instruction sets like AVX so this is a good thing to think about.

**Comment by blaiseli** [comment]
@yesimon About portability: I thought that the the situation was either one of the following:
1. The original program is available as binary.
2. The original program is only available as source code.

In situation 1, conda has two possibilities:
1.1 Chose a rather "universal" binary.
1.2 Find the source and provide a recipe for compiling it.

For situation 2, conda does the same as 1.2.

I thought that in no case did conda compile from source and distribute a conda-compiled binary. When compilation happens, it occurs on the package installation machine, isn't it?
That's why I don't understand why a recipe with `-march=native` would be problematic. This option will make the compilation use instructions that are available on the machine. I thought that this was the best way to ensure the binary is compatible with the machine on which it will run: compile the program with the same processor that will run it.

Did I miss something?
Does `-march=native` sometimes cause compilation to fail?
Or does conda actually sometimes compile and distribute binaries? (If I understand correctly, a compilation may happen in a docker image on the recipe maintainer's machine, but this is just for testing purpose, the resulting binary is not what is distributed.)

In situation 1.1, no need to give a compiling recipe, so there is no `-march` option to chose whatsoever.

Regarding Gentoo: I happen to have a rather old laptop as personal home computer (a 2007 model), and I use Gentoo on it. Compiling from source is actually not difficult in this case, because Gentoo maintainers provide recipes. I don't have to figure it out by myself. It indeed takes much more time to install or update programs than with other distros, and I made no test to verify that programs run faster than if I had them installed from binary packages, but the computer runs smoothly enough that I don't want to buy another machine yet.

**Comment by blaiseli** [comment]
After some further reading and some experimentation (installing packages with conda takes much less time than with my install scripts that compile them from source), I understand now that conda actually does install a pre-compiled binary when possible, and that the build recipe is used for package creation and is not meant to be run on the user's machine. Sorry for my misunderstanding.

It would be nice to have an option to force package build when installing at the user's machine, with the possibility to apply custom flags such as `-march=native`.

**Comment by yhoogstrate** [comment]
For those who are curious about the possible impact of adding the `-march=native` compile flag, you can find frequently benchmarks at Phoronix.

Here you have an example: http://www.phoronix.com/scan.php?page=news_item&px=GCC-6.1-Compiler-Optimizations

Of course, it's architecture, written code and GCC version dependent.

**Comment by kyleabeauchamp** [comment]
I wanted to check in on the status of two questions:

1.  What is the baseline "assumed" instruction set available---SSE2?  
2.  For more recent instruction sets, are "features" or "recipe copies" the preferred mechanism for handling permutations of builds with more recent instruction sets?

PS I've asked the same question on conda-forge (https://github.com/conda-forge/conda-forge.github.io/issues/49) to inform our discussion here.

**Comment by kyleabeauchamp** [comment]
Cross-referencing one other ticket (https://github.com/bioconda/bioconda-recipes/issues/5297), which was ambiguous as to whether SSE3 was considered "available by default".

**Comment by kyleabeauchamp** [comment]
The one issue I have is that the `variants` mechanism seems like it might be somewhat brittle, at least when compared to an approach of manually duplicating the recipe.  We'll have to investigate this with some experiments once CB3 goes live.
